### PR TITLE
VM Copy support

### LINKED ISSUE #562: vcd_vapp_vm - Clone from running VM resource
Hi,

We have a scenario where in vCloud Director we need to create new VMs from a running VM, rather than a Template VM in a catalog library.
I can't find an option with Terraform to use the vCloud director "Copy To" function. Is this available in Terraform at present, or if not possible at all? If it is possible but not available yet, could I raise a feature request please?

Many thanks.

### PR DESCRIPTION
**Pull request by Didainius**
This PR adds support for "VM Copy" operation. This allows users to create VMs from from existing VMs. This works by adding `copy_from_vm_id` in `vcd_vapp_vm` and `vcd_vm` resources.

There is one caveat due to a known VCD bug - when creating a Standalone VM copy and source VM is in different VDC, - destination VM must have a `sizing_policy_id` specified, as otherwise it will return an error (this is also highlighted in docs)

Flow diagram for `create` functions of `vcd_vapp_vm` and `vcd_vm` (updated from PR #901 version to reflect latest changes)
* Purple shows the main flow of code for `Create`
* Green blocks signify API call definition for each of VM types that are being created.
* Yellow blocks signify where additional code can be hooked up for particular scope

```mermaid
flowchart TD
    S[VM Create]
    S ==> AAAA
    AAAA{Standanlone or vApp VM?}
    AAAA --> |resource vcd_vm| AA{vApp VM or Standalone}
    AAAA --> |resource vcd_vapp_vm| A{vApp VM or Standalone}

    AA[func resourceVcdStandaloneVmCreate] --> B[func genericResourceVmCreate]
    A[func resourceVcdVAppVmCreate] -->|lock parent vApp\n defer unlock| B[func genericResourceVmCreate]

    BB{Is template, empty, or VM Copy}

    B --> BB

    BB -->|"isVmFromTemplate || isVmCopy"| D[func createVmFromImage]
    D --> D3[func lookupSourceImage]
    D3 -->|"isVmFromTemplate"| D4[func lookupSourceCatalogTemplate]
    D3 -->|"isVmCopy"| D5[func lookupSourceVm]

    D4 --> D7[Source Image Identified]
    D5 --> D7

    BB -->|isEmptyVm| E[func createVmEmpty]

    DDD([vApp VM from template])
    DDDD(Standalone VM from template)
    

    EEE(Empty vApp VM)
    EEEE(Empty Standalone VM)

    DD([Optionally perform template VM specific actions])
    EE([Optionally perform empty VM specific actions])


   DDOPTIONAL1([Optionally perform vApp VM template specific actions])
   DDOPTIONAL2([Optionally perform Standalone VM template specific actions])

   EEOPTIONAL1([Optionally perform Empty vApp VM specific actions])
   EEOPTIONAL2([Optionally perform Empty Standalone VM specific actions])


    D7 --> |Is vApp VM| DDD
    D7 --> |Is Standalone VM| DDDD

    E --> |Is vApp VM| EEE
    E --> |Is Standalone VM| EEEE

    DDD --> DDOPTIONAL1
    DDOPTIONAL1 --> DD
    DDDD --> DDOPTIONAL2
    DDOPTIONAL2 --> DD

    EEE --> EEOPTIONAL1
    EEOPTIONAL1 --> EE
    EEEE --> EEOPTIONAL2
    EEOPTIONAL2 --> EE


    F[func genericResourceVmCreate \n'Common API calls for ALL VMs' ]
    DD --> F
    EE --> F


    F ==> G
    G ==> Z
    
    G[func genericResourceVmCreate\nlast step -> power on]

    Z[VM Created]

style DDD fill:#45f248,stroke:#333,stroke-width:4px
style DDDD fill:#45f248,stroke:#333,stroke-width:4px
style EEE fill:#45f248,stroke:#333,stroke-width:4px
style EEEE fill:#45f248,stroke:#333,stroke-width:4px


style F fill:#F1EE8E,stroke:#333,stroke-width:4px
style DD fill:#F1EE8E,stroke:#333,stroke-width:4px
style EE fill:#F1EE8E,stroke:#333,stroke-width:4px
style DDOPTIONAL1 fill:#F1EE8E,stroke:#333,stroke-width:4px
style DDOPTIONAL2 fill:#F1EE8E,stroke:#333,stroke-width:4px

style EEOPTIONAL1 fill:#F1EE8E,stroke:#333,stroke-width:4px
style EEOPTIONAL2 fill:#F1EE8E,stroke:#333,stroke-width:4px
```

Closes #562 

Acceptance tests passed in 10.4.0 and 10.5.1

### CODE REVIEW

File: vcd/resource_vcd_vapp_vm.go [hunk 1]
```diff
@@ -967,13 +989,13 @@ func genericResourceVmCreate(d *schema.ResourceData, meta interface{}, vmType ty
...
 
-// createVmFromTemplate is responsible for create VMs from template of two types:
+// createVmFromImage is responsible for create VMs from template of two types:
 // * Standalone VMs
 // * VMs inside vApp (vApp VMs)
 //
 // Code flow has 3 layers:
 // 1. Lookup common information, required for both types of VMs (Standalone and vApp child). Things such as
-//   - Template to be used
+//   - VM Source Image - Catalog Template or VM (VM Copy operation) to be used
```
Comment (lvirbalas) [inline vcd/resource_vcd_vapp_vm.go:998]: Maybe this for more clarity:

```suggestion
//   - VM Source Image - Catalog Template or VM (for VM Copy operation) to be used
```

**Review by lvirbalas — APPROVED** [review]
This is great! Just a few suggestions inline.

File: vcd/resource_vcd_vapp_vm_tools.go [hunk 2]
```diff
@@ -23,6 +23,45 @@ import (
 	"github.com/vmware/go-vcloud-director/v2/util"
 )
 
+// getVmSourceImage retrieves non-empty VM source image reference. It can be one of:
+// * Catalog VM template (regular way for creating VMs)
+// * Already running VM templates (VM Copy). The VM must be within the same Org
```
[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/resource_vcd_vapp_vm_tools.go]: Fixed.

File: vcd/resource_vcd_vapp_vm.go (hunk 1 above)
Comment (Didainius) [reply vcd/resource_vcd_vapp_vm.go]: Fixed.

File: website/docs/r/vapp_vm.html.markdown [hunk 3]
```diff
@@ -442,7 +442,36 @@ resource "vcd_vapp_vm" "advancedVM" {
   cpu_reservation = "200"
   cpu_limit       = "1000"
 }
+```
+
+## Example Usage (VM copy)
+This example shows how to create a VM copy from existing VM
```
Comment (dataclouder) [inline website/docs/r/vapp_vm.html.markdown:448]:
```suggestion
This example shows how to create a copy of an existing VM
```

File: website/docs/r/vapp_vm.html.markdown [hunk 4]
```diff
@@ -442,7 +442,36 @@ resource "vcd_vapp_vm" "advancedVM" {
...
+  vapp_name = data.vcd_vapp.web.name
+  name      = "web1"
+}
+
+resource "vcd_vapp_vm" "vm-copy" {
+  org = "org"
+  vdc = "vdc"
 
+  copy_from_vm_id = data.vcd_vapp_vm.existing.id # source VM ID
+  vapp_name       = vcd_vapp.vm-copy-destination-template-vm.name
```
Comment (dataclouder) [inline website/docs/r/vapp_vm.html.markdown:462]:
```suggestion
  vapp_name       = data.vcd_vapp_vm.existing.vapp_name
```

File: website/docs/r/vapp_vm.html.markdown [hunk 5]
```diff
@@ -442,7 +442,36 @@ resource "vcd_vapp_vm" "advancedVM" {
...
+  name            = "VM Copy"
+  power_on        = false
+
+  network {
+    type               = "org"
+    name               = vcd_vapp_org_network.template-vapp-copy.org_network_name
```
Comment (dataclouder) [inline website/docs/r/vapp_vm.html.markdown:468]: Where does this vApp network come from?

File: website/docs/r/vapp_vm.html.markdown [hunk 6]
```diff
@@ -456,6 +485,10 @@ The following arguments are supported:
 * `computer_name` - (Optional; *v2.5+*) Computer name to assign to this virtual machine.
 * `vapp_template_id` - (Optional; *v3.8+*) The URN of the vApp Template to use. You can fetch it using a [`vcd_catalog_vapp_template`](/providers/vmware/vcd/latest/docs/data-sources/catalog_vapp_template) data source.
 * `vm_name_in_template` - (Optional; *v2.9+*) The name of the VM in vApp Template to use. For cases when vApp template has more than one VM.
+* `copy_from_vm_id` - (Optional; *v3.12+*) The ID of existing VM to make a copy of it. The source VM
```
Comment (dataclouder) [inline website/docs/r/vapp_vm.html.markdown:488]: We should specify that when using this option we can't use reference to a vApp template

File: website/docs/r/vapp_vm.html.markdown (hunk 3 above)
Comment (Didainius) [reply website/docs/r/vapp_vm.html.markdown]: Fixed.

File: website/docs/r/vapp_vm.html.markdown (hunk 4 above)
Comment (Didainius) [reply website/docs/r/vapp_vm.html.markdown]: Fixed.

File: website/docs/r/vapp_vm.html.markdown (hunk 5 above)
Comment (Didainius) [reply website/docs/r/vapp_vm.html.markdown]: I obviously didn't have a data source for it, but changed now - added additional data source.

File: website/docs/r/vapp_vm.html.markdown (hunk 6 above)
Comment (Didainius) [reply website/docs/r/vapp_vm.html.markdown]: Fixed.

**Review by dataclouder — APPROVED** [review]

**Review by adambarreiro — APPROVED** [review]
LGTM!

**Pull request merged**